src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryTaskLauncher.java

```diff
@@ -0,0 +1,607 @@
+import org.cloudfoundry.client.v3.packages.Package;
+import org.cloudfoundry.client.v3.packages.PackageType;
+import org.cloudfoundry.client.v3.packages.StagePackageRequest;
+import org.cloudfoundry.client.v3.packages.State;
+import org.cloudfoundry.client.v3.packages.UploadPackageRequest;
+import org.cloudfoundry.client.v3.servicebindings.CreateServiceBindingRequest;
+import org.cloudfoundry.client.v3.servicebindings.Relationships;
+import org.cloudfoundry.client.v3.servicebindings.ServiceBindingType;
+import org.cloudfoundry.client.v3.tasks.CancelTaskRequest;
+import org.cloudfoundry.client.v3.tasks.CancelTaskResponse;
+import org.cloudfoundry.client.v3.tasks.CreateTaskRequest;
+import org.cloudfoundry.client.v3.tasks.CreateTaskResponse;
+import org.cloudfoundry.client.v3.tasks.GetTaskRequest;
+import org.cloudfoundry.client.v3.tasks.GetTaskResponse;
+import org.cloudfoundry.client.v3.tasks.Task;
+import org.cloudfoundry.operations.CloudFoundryOperations;
+import org.cloudfoundry.operations.services.ServiceInstance;
+import org.cloudfoundry.util.PaginationUtils;
+import org.cloudfoundry.util.ResourceUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.deployer.spi.core.AppDeploymentRequest;
+import org.springframework.cloud.deployer.spi.task.LaunchState;
+import org.springframework.cloud.deployer.spi.task.TaskLauncher;
+import org.springframework.cloud.deployer.spi.task.TaskStatus;
+import org.springframework.util.CollectionUtils;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.lang.Integer.parseInt;
+import static java.lang.String.valueOf;
+import static org.cloudfoundry.util.DelayUtils.exponentialBackOff;
+import static org.cloudfoundry.util.tuple.TupleUtils.function;
+import static org.springframework.util.StringUtils.commaDelimitedListToSet;
+
+/**
+ * {@link TaskLauncher} implementation for CloudFoundry.  When a task is launched, if it has not previously been
+ * deployed, the app is created, the package is uploaded, and the droplet is created before launching the actual
+ * task.  If the app has been deployed previously, the app/package/droplet is reused and a new task is created.
+ *
+ * @author Greg Turnquist
+ * @author Michael Minella
+ */
+public class CloudFoundryTaskLauncher implements TaskLauncher {
+
+    private static final Logger logger = LoggerFactory
+        .getLogger(CloudFoundryTaskLauncher.class);
+
+    private final CloudFoundryClient client;
+
+    private final CloudFoundryOperations operations;
+
+    private final CloudFoundryDeployerProperties properties;
+
+    public static final String SERVICES_PROPERTY_KEY = "spring.cloud.deployer.cloudfoundry.defaults.services";
```

Comment:
Should remove `.defaults.` from here. Actually, should extract / reuse the ones from `CloudFoundryAppDeployer` I guess